### PR TITLE
Generalize (GMonoid (Maybe a)) to require only (GSemigroup a)

### PR DESCRIPTION
**Pull request by sjakobi**

### CODE REVIEW

**Comment by sjakobi** [comment]
I had to move the instance to `Generics.Deriving.Semigroup` because it already depends on `G.D.Monoid` for the `WrappedMonoid` instance.

Now of course we get 

```
src/Generics/Deriving/Semigroup.hs:149:1: warning: [-Worphans]
    Orphan instance: instance [safe] GSemigroup a => GMonoid (Maybe a)
    To avoid this
        move the instance declaration to the module of the class or of the type, or
        wrap the type with a newtype and declare the instance on the new type.
    |
149 | instance GSemigroup a => GMonoid (Maybe a) where
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

Not sure about the best way to solve this.

**Comment by RyanGlScott** [comment]
Here is my recommendation:

1. Move the current contents of `Generics.Deriving.Monoid` into a new `Generics.Deriving.Monoid.Internal` module, which isn't exposed publicly.
2. Change `Generics.Deriving.Semigroup` to import `Generics.Deriving.Monoid.Internal`, but keep everything else the same.
3. In the new `Generics.Deriving.Monoid` module, define the `instance GSemigroup a => GMonoid (Maybe a)` that you seek, and reexport the rest of `Generics.Deriving.Monoid.Internal`. This instance will technically be an orphan, so you'll have to enable `{-# OPTIONS_GHC -fno-warn-orphans #-}` in that module to suppress the resulting warning, but this should be fine.

**Comment by sjakobi** [comment]
Thanks Ryan! That was a very good plan! :+1:

File: src/Generics/Deriving/Monoid.hs [hunk 1]
```diff
@@ -1,289 +1,15 @@
...
-  gmappend' :: f x -> f x -> f x
-
-instance GMonoid' U1 where
-  gmempty' = U1
-  gmappend' U1 U1 = U1
-
-instance GMonoid a => GMonoid' (K1 i a) where
-  gmempty' = K1 gmempty
-  gmappend' (K1 x) (K1 y) = K1 (x `gmappend` y)
+module Generics.Deriving.Monoid (module Generics.Deriving.Monoid.Internal) where
```
[the remark this answers is not in the document]
Comment (sjakobi) [reply src/Generics/Deriving/Monoid.hs]: I imitated `D.G.Base.Internal` and simply transformed the module header into a docs chunk in `D.G.Monoid.Internal`'s export list.

The haddocks for `D.G.Monoid` start like this now: 
image

File: src/Generics/Deriving/Monoid.hs [hunk 2]
```diff
@@ -1,289 +1,15 @@
...
 #if __GLASGOW_HASKELL__ >= 701
-{-# LANGUAGE DefaultSignatures #-}
-#endif
-
-#if __GLASGOW_HASKELL__ >= 705
-{-# LANGUAGE PolyKinds #-}
-#endif
-
-#if __GLASGOW_HASKELL__ >= 710
 {-# LANGUAGE Safe #-}
```
[the remark this answers is not in the document]
Comment (sjakobi) [reply src/Generics/Deriving/Monoid.hs]: Fixed!

Note that `G.D.{Base,ConNames,Default}` all have the following setup though:

```
#if __GLASGOW_HASKELL__ >= 701
{-# LANGUAGE Safe #-}
#endif
```

[the remark this answers is not in the document]
Comment (RyanGlScott) [reply src/Generics/Deriving/Monoid.hs]: True. I've opened #65 about this issue.

File: src/Generics/Deriving/Monoid.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply src/Generics/Deriving/Monoid.hs]: I see, that works just as well.

**Pull request merged**